File: ui/src/lib/api/security.ts
import { api } from '../fetcher';

export interface AuthProviderItem {
  id: 'local' | 'oidc' | 'ldap' | 'saml';
  name: string;
  description: string;
  enabled: boolean;
  config_json: string;
  updated_at: string;
}

export interface TwoFactorStatus {
  enabled: boolean;
  configured: boolean;
  configured_at: string;
}

export interface SecurityOverview {
  auth_providers: AuthProviderItem[];
  two_factor: TwoFactorStatus;
  admin_username: string;
  password_last_updated: string;
}

export interface Init2FAOutput {
  secret: string;
  otpauth_url: string;
  qr_svg: string;
}

export interface Verify2FAOutput {
  enabled: boolean;
  recovery_codes: string[];
}

export const securityApi = {
  getOverview: () => api.get<SecurityOverview>('/api/v1/settings/security'),
  updateProvider: (id: string, enabled: boolean, configJSON: string) =>
    api.put<{ message: string; id: string; enabled: boolean }>(`/api/v1/settings/security/auth-providers/${id}`, {
      enabled,
      config_json: configJSON,
    }),
  init2FA: () => api.post<Init2FAOutput>('/api/v1/settings/security/2fa/init', {}),
  verify2FA: (secret: string, code: string) =>
    api.post<Verify2FAOutput>('/api/v1/settings/security/2fa/verify', { secret, code }),
  disable2FA: () => api.post<{ message: string }>('/api/v1/settings/security/2fa/disable', {}),
  changePassword: (currentPassword: string, newPassword: string) =>
    api.post<{ message: string }>('/api/v1/settings/security/change-password', {
      current_password: currentPassword,
      new_password: newPassword,
    }),
};

File: control-plane/internal/domain/entity/security.go
package entity

// Tuân thủ Flat Entity: không chứa json tags.

// AuthProviderItem đại diện cho một phương thức xác thực người dùng trong hệ thống:
// local, oidc, ldap, saml.
type AuthProviderItem struct {
	ID          string
	Name        string
	Description string
	Enabled     bool
	ConfigJSON  string
	UpdatedAt   string
}

// TwoFactorStatus phản ánh trạng thái kích hoạt 2FA của tài khoản người dùng hiện tại.
type TwoFactorStatus struct {
	Enabled      bool
	Configured   bool
	ConfiguredAt string
}

// SecurityOverview là flat projection toàn cảnh trang Security Settings.
type SecurityOverview struct {
	AuthProviders       []AuthProviderItem
	TwoFactor           TwoFactorStatus
	AdminUsername       string
	PasswordLastUpdated string
}

// Init2FAOutput chứa thông tin cấp mới cho người dùng để quét mã QR và lưu secret.
type Init2FAOutput struct {
	Secret     string
	OtpAuthURL string
	QrSVG      string
}

// Verify2FACommand mang dữ liệu mã xác thực người dùng gửi lên để kích hoạt 2FA.
type Verify2FACommand struct {
	Secret string
	Code   string
}

// Verify2FAOutput trả về kết quả kích hoạt cùng danh sách mã khôi phục khẩn cấp.
type Verify2FAOutput struct {
	Enabled       bool
	RecoveryCodes []string
}

// ChangePasswordCommand nhận mật khẩu hiện tại và mật khẩu mới để đổi thông tin đăng nhập.
type ChangePasswordCommand struct {
	CurrentPassword string
	NewPassword     string
}
